### PR TITLE
Latency/jitter improvements for control loop

### PR DESCRIPTION
**Pull request by borisbstyle**
First step to some task scheduling. Especially usefull with gyro_sync

For the moment just processRx and reduced latency in acro mode

Baro would also require same as it can cause up to 20% jitter in combination with gyro_sync

### CODE REVIEW

File: src/main/flight/imu.c [hunk 1]
```diff
@@ -297,12 +297,13 @@ static void imuCalculateEstimatedAttitude(void)
...
 }
 
-void imuUpdate(rollAndPitchTrims_t *accelerometerTrims)
+void imuUpdate(rollAndPitchTrims_t *accelerometerTrims, uint8_t imuUpdateSensors)
 {
-    gyroUpdate();
-
-    if (sensors(SENSOR_ACC)) {
-        updateAccelerationReadings(accelerometerTrims); // TODO rename to accelerometerUpdate and rename many other 'Acceleration' references to be 'Accelerometer'
+    if (imuUpdateSensors == ONLY_GYRO || imuUpdateSensors == ACC_AND_GYRO) {
```
Comment (ledvinap) [inline src/main/flight/imu.c]: using bit for each sensor will probably result in cleaner code. You can even reuse SENSOR_* codes

**Comment by ledvinap** [comment]
processRX handling is IMO step in wrong direction. 

The problem is quite complex, in the end you will be reinventing RTOS. It will be very hard to find all bugs.
Either find some KISS approach (maybe running rx once per loop just after update is enough), or port CF to some RTOS (possibly using just  'slow' and 'fast' task and keeping most of current task planning code)

**Comment by borisbstyle** [comment]
@ledvinap 
I didnt exactly check what part of the baro was the slow one. I just profiled all functions with their jitter contribution and those 2 came on the top.

I rebased and implemented single task run before the loop. What do you think of this approach?

**Comment by ledvinap** [comment]
It may be easier to run tasks just after processing (in looptime `if`), without waiting for next loop. But that is implementation detail and waiting may spread processing slightly better.

Bigger problem is that RC processing is now called only on 1/size(periodicTasks) rate. This may be too low ... 

Ideally as many tasks as possible will be run before next looptime (running each only once). And if some tasks are left, they should run on next loop (taking care to prevent task starvation).

I'am afraid it will be quite complex problem ...

**Comment by digitalentity** [comment]
@borisbstyle, lets make an RTOS-like dynamic scheduler that would measure task execution time and schedule them appropriately. Gyro processing will have the highest priority.

**Comment by digitalentity** [comment]
@borisbstyle, I have some task scheduling in the works at the moment. Will post a proof of concept commit for review in a few hours.

**Comment by digitalentity** [comment]
I've made a quick and dirty implementation of cooperative task scheduler here 29bfb71717ef54517e6ac3b9ac3e3baa5bdf390e. Commit is based on inav-dev branch, but the code should me mergeable with CF with little effort.

A pleasent surprise to me was that code footprint got ~400 bytes smaller.

Some real CLI data with new scheduling (CC3D, MW23):
```
# status
System Uptime: 199 seconds, Voltage: 0 * 0.1V (3S battery - OK), CPU load: 30%
CPU Clock=72MHz, GYRO=MPU6000, ACC=MPU6000
Cycle Time: 2003, I2C Errors: 0, config size: 1804

# tasks
Task list:
0 - GYRO/PID, time = 61095 ms
1 - ACCEL, time = 2616 ms
2 - RX, time = 694 ms
3 - GPS, time = 23 ms
4 - COMPASS, time = 23 ms
5 - BARO, time = 45 ms
6 - SONAR, time = 45 ms
7 - DISPLAY, time = 22 ms
8 - TELEMETRY, time = 548 ms
```

**Comment by digitalentity** [comment]
New scheduler has some simple priority management and is able to keep up with 700us looptime on CC3D with acc enabled and CPU load of 83% with minimal jitter (~7-8 us)

**Comment by borisbstyle** [comment]
That's great! I was actually just doing some research on different task
scheduler approaches, but thanks this will speed it all up.

I am going to check it and try to test/review it this evening on various
targets.
Op 19 nov. 2015 08:22 schreef "Konstantin Sharlaimov" <
notifications@github.com>:

**Comment by digitalentity** [comment]
This is just a quick hack, benchtested, never flown. Lots of work to be done until this scheduler is reliable and ready. To me the code looks cleaner and simpler now - previous `loop()` was a mess.

**Comment by digitalentity** [comment]
I didn't go for full blown preemptive scheduler as it is going to be an overkill and would actually introduce more jutter. I tried to combine the simplicity of CF's superloop approach with ability to rob the other tasks' time in favor of gyro/pid processing and different execution rated for different tasks.

**Comment by borisbstyle** [comment]
This is a great step.

CPU usage is also very usefull as we have so many targets, but now we can
compare them properly.

Once I get some time later today I will start testing.
Op 19 nov. 2015 08:40 schreef "Konstantin Sharlaimov" <
notifications@github.com>:

**Comment by ledvinap** [comment]
@digitalentity : Can you please modify the code to track max task execution time? It would be very helpful for jitter estimation

**Comment by ledvinap** [comment]
BTW: Here is my priority queue based timing implementations. It is interrupt based, but similar approach may be used to plan loop()-level functions

**Comment by ledvinap** [comment]
https://github.com/ledvinap/cleanflight/blob/local/src/main/drivers/timer_queue.c

**Comment by digitalentity** [comment]
@ledvinap done. Moved some tasks out from `annexCode`, cycleTime is now even more stable:
```
# status
System Uptime: 48 seconds, Voltage: 0 * 0.1V (3S battery - OK), CPU load: 43%
CPU Clock=72MHz, GYRO=MPU6000, ACC=MPU6000, BARO=MS5611
Cycle Time: 2021, I2C Errors: 0, config size: 1804

# tasks
Task list:
0 - GYRO/PID, max time = 1291 us, total time = 20133 ms
1 - ACCEL, max time = 178 us, total time = 658 ms
2 - SERIAL, max time = 16883 us, total time = 821 ms
3 - BEEPER, max time = 32 us, total time = 72 ms
4 - RX, max time = 165 us, total time = 229 ms
5 - GPS, max time = 28 us, total time = 7 ms
6 - COMPASS, max time = 28 us, total time = 7 ms
7 - BARO, max time = 266 us, total time = 161 ms
8 - SONAR, max time = 53 us, total time = 20 ms
9 - DISPLAY, max time = 28 us, total time = 6 ms
10 - TELEMETRY, max time = 35 us, total time = 217 ms
```

**Comment by digitalentity** [comment]
And the same code on my NAZE32 rev5 full:
```
# status
System Uptime: 23 seconds, Voltage: 0 * 0.1V (1S battery - NOT PRESENT), CPU load: 27%
CPU Clock=72MHz, GYRO=MPU6050, ACC=MPU6050.n, BARO=MS5611, MAG=HMC5883
Cycle Time: 3510, I2C Errors: 0, config size: 2004

# tasks
Task list:
0 - GYRO/PID, max time = 1135 us, total time = 9259 ms
1 - ACCEL, max time = 240 us, total time = 727 ms
2 - SERIAL, max time = 18958 us, total time = 438 ms
3 - BEEPER, max time = 18 us, total time = 43 ms
4 - RX, max time = 73 us, total time = 109 ms
5 - GPS, max time = 16 us, total time = 4 ms
6 - COMPASS, max time = 240 us, total time = 76 ms
7 - BARO, max time = 241 us, total time = 105 ms
8 - SONAR, max time = 15 us, total time = 8 ms
9 - DISPLAY, max time = 15 us, total time = 4 ms
10 - TELEMETRY, max time = 19 us, total time = 128 ms
11 - LEDSTRIP, max time = 16 us, total time = 42 ms
```

Note how much longer ACCEL task is taking. CC3D has accel on SPI bus while NAZE uses I2C. This gives almost 60us difference in ACCEL processing.

**Comment by digitalentity** [comment]
How the simplest reordering of the code execution can save about 1.3 kB FLASH on my iNav branch?

Before the scheduler:
```
   text    data     bss     dec     hex filename
 127012     284   15868  143164   22f3c ./obj/main/cleanflight_NAZE.elf
```

After the scheduler:
```
   text    data     bss     dec     hex filename
 125636     696   15588  141920   22a60 ./obj/main/cleanflight_NAZE.elf
```

**Comment by digitalentity** [comment]
@borisbstyle, @ledvinap what do you think? Should I apply the scheduler patches to the CF's main and post a PR?

**Comment by henn1001** [comment]
Sounds & looks great guys!

**Comment by ProDrone** [comment]
@digitalentity 
Yes, looking great, make a PR!

**Comment by digitalentity** [comment]
Ok, will post a PR in a few

**Comment by ledvinap** [comment]
Very  interesting

> 2 - SERIAL, max time = 18958 us, total time = 438 ms 

this is probably some glitch, 20ms is quite a lot of time (maybe some initialization? This should definitely not happen when armed).

With this simple scheduler approach worst jitter will be equal to longest running task. That is ~250us (baro, accel).

With gyro sync, it should be possible to lock to gyro sample rate and delay ESC output to get perfect jitter-free operation.

**Comment by digitalentity** [comment]
@ledvinap 
The long running serial task is probably CLI processing or some initialization. Scheduler can be modified to account for average task execution time and delay its execution until a sufficient interval of free time is available. The other question is do we need it?

**Comment by ledvinap** [comment]
One improvement may be adding oneshot-like feature. It can be eventually used by tasks that wait some time (baro, etc), simplifying the code further..

**Comment by digitalentity** [comment]
PR #1504 created.

**Comment by digitalentity** [comment]
@ledvinap not sure I understand. Can you please elaborate a little?

**Comment by ledvinap** [comment]
there are lots of places like https://github.com/cleanflight/cleanflight/blob/master/src/main/sensors/barometer.c#L129

Allowing task to plan self to run once after delay will make it unnecessary ..

**Comment by digitalentity** [comment]
@ledvinap, that's a good one! Lets get the simplest implementation working and merged first and improve that later? :smile:

**Comment by digitalentity** [comment]
@ledvinap,
Actually, new scheduler already has the `setTaskDesiredPeriod` function to set desired execution period. Currently it is used to setup the Gyro/PID task to execute at `looptime` intervals.
The task can reschedule itself on the fly using this API.

**Comment by digitalentity** [comment]
@ledvinap, done. See https://github.com/cleanflight/cleanflight/blob/42c630ca5886c04bd4ba84616fa36648a748521c/src/main/mw.c#L793 and https://github.com/cleanflight/cleanflight/blob/42c630ca5886c04bd4ba84616fa36648a748521c/src/main/sensors/barometer.c#L124

**Comment by digitalentity** [comment]
Same logic can probably be implemented for telemetry. GPS and beeper. No need to call the task to figure out the call was too early.

**Comment by ledvinap** [comment]
@digitalentity : This is the delay reference time bug again - the scheduler calculates delay from `lastExecutedAt`, not time when `rescheduleTask` was called. Time between lastExecutedAt and rescheduleTask is subtracted from delay. Resulting delay may be too short for external hardware to finish processing.

**Comment by digitalentity** [comment]
True, thanks. Will fix.

**Comment by ledvinap** [comment]
This is big reason to use mature RTOS - similar bugs are already fixed ...

**Comment by digitalentity** [comment]
@ledvinap Real RTOS have other issues like large task switching overhead leading to a relatively big time quant leading to jitter or lower PID update frequency. Superloop approach is faster and easier to debug.

**Comment by borisbstyle** [comment]
OP projects also faced simillar overhead problems with RTOS from what I remember.

We can test both approaches though

**Comment by digitalentity** [comment]
My approach doesn't add overhead of context switching, the downside is that it's up to coder to properly break and schedule lengthy tasks so they don't clog the scheduler.

**Comment by ledvinap** [comment]
Yes, I can imagine the trade-offs ...

Task switch time can be around  100 instructions/1.5us (freeRTOS/ChibiOS), probably not that important (unless things like softserial are considered).

RAM is quite plentiful in CF, FLASH may be a problem (but some functions will be much simpler, it may well offset RTOS penalty)

Concurrent programming is big problem. Even CF has concurrency problems  (rcData overwrite, etc). Properly synchronizing access to all variables will need lots of testing.

The superloop approach is great as long as all operations are run on each loop. Scheduling will introduce data dependency problems, task ordering etc.  

I'am not sure what is the 'right' way (and @hydra is in charge ... ). New scheduler seems like step in right direction - just use abstract interface and try to hide implementation details, so the plumbing may be easily changed later ...

**Comment by borisbstyle** [comment]
We need to continue the discussion on the right place. This is the wrong PR now
https://github.com/cleanflight/cleanflight/pull/1504

**Pull request closed without merging**